### DISCUSSION TITLE
[Question]<How to use ESP3D with the MCP23017 component in the components/expander/io_expander/mcp23017 directory?>

### DISCUSSION THREAD

**Discussion by liu2-3zhi** [luc-github/ESP3D#947] [Questions]
How to use ESP3D with the MCP23017 component in the components/expander/io_expander/mcp23017 directory? https://github.com/espressif/esp-iot-solution/tree/d7f413d058e7acbc1832305c65d211ee13fd952d/components/expander/io_expander/mcp23017

**Comment by liu2-3zhi** [comment]
Please forgive my desire, I want to enable all features, including Ethernet, camera, buzzer, and so on. The number of I/O has prevented this behavior.

**Comment by luc-github** [comment]
I am not sure because I never needed to code ioexpander but for arduino I guess you can use this arduino library https://www.arduino.cc/reference/en/libraries/mcp23017/

**Comment by liu2-3zhi** [comment]
Please forgive my stupidity, I don't know how to define pin.

After add 
```c
#include "mcp23017.h"
#define I2C_MASTER_SCL_IO 21 /*!< gpio number for I2C master clock IO21*/
#define I2C_MASTER_SDA_IO 22 /*!< gpio number for I2C master data IO15*/
#define I2C_MASTER_NUM I2C_NUM_1 /*!< I2C port number for master dev */
#define I2C_MASTER_TX_BUF_DISABLE 0 /*!< I2C master do not need buffer */
#define I2C_MASTER_RX_BUF_DISABLE 0 /*!< I2C master do not need buffer */
#define I2C_MASTER_FREQ_HZ 100000 /*!< I2C master clock frequency */ 
```
For example: 
```c
#define OLED_PIN_SDA 4
```
Should I change it directly to 
```c
#define OLED_PIN_SCL MCP23017_PIN0
```
or not?

**Reply by luc-github** [reply]
I never used directly an io expander but it is better that your define  do not use already existing pin number, to avoid any confusion , so IOExpander usually start at 100
so when pin number is >=100 you know it must be used by IO expander and so use the correct function to use it , of course need adjust the GPIO value like 100 to MCP23017_PIN0, etc...
